### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by trthatcher** [trthatcher/MLKernels.jl#12]
@st-- 

I think we should establish a roadmap. For the most part, I already have what I need from this package. All I need to do is expand/finish the approximations.

We have a good foundation for the kernel derivatives defined for the Gaussian kernel. I can expand it to include all the other kernels, it's just a matter of sinking some time into it. 

What are you priorities in terms of the features? How do you want to approach it?

**Comment by st--** [comment]
Well, when I came across your package I thought it'd be better to focus on a single, well-done package rather than keeping rolling my own. This I already have in my own (and need to port to MLKernels.jl before I can get rid of my covariances.jl):
- Kernels
  - [ ] Automatic Relevance Determination (ARD) kernels: scale ||x-y|| or dot(x,y) with a weights array
    - [ ] derivatives w.r.t. the weights (parameter array)
  - [ ] composites of composite kernels
    - [ ] derivatives w.r.t. parameters of subkernels
  - [ ] periodic kernel: `exp(-2sin(π*(x-y)/k.p)^2 / k.ell^2)`
- Methods
  - [ ] optimized matrices of kernel derivatives (need to be careful about memory allocation; that made my first stab at it very slow...)

Not a priority for myself, but should be implemented before we can call it "completed":
- [ ] expand derivatives to the other kernels

**Comment by trthatcher** [comment]
Okay, I added a [wiki](https://github.com/trthatcher/MLKernels.jl/wiki). We can organize there.

I was going to have some free time to spend on development the next couple days. Can you mark what you'll be working on and, if you'd like, mark anything that's important to you feature-wise that you'd like me to work on. That way we can prioritize and we won't have any overlap in work (like there was with the most recent commit)

**Comment by st--** [comment]
I'll be away until Tuesday. (I've uploaded my WIP on kernel matrix derivatives. The in-place version now performs almost as well as my own-brewn previous function.) I'd quite like to figure out how to implement all the ARD stuff myself (though do say if you come up with good ideas), and on the other hand if you could implement the weighted gramian calculation that'd be helpful. Everything else, if you get to it, great, and if not, I can have a go at it once I'm back!

**Comment by trthatcher** [comment]
I played around with the kernel derivatives and cleaned up code. 

Just a note, I'd like to keep the kernelize function scalar. It should be a function that can be inlined easily. As such, it should take a scalar and return a scalar. The same for any of its derivatives.

Kernel matrices are usually a simple, element-wise transformations of another matrix. The element level transformation is kernelize and the 'another matrix' is usually a gramian of some kind.

I've changed the names of some things for simplicity (partially noted in the style guide). Generally speaking, we're looking at kernels of a form k(x,y) = kappa(z(x,y)) where z maps two vectors to a scalar. kappa is the old kernelize function. z is either sqdist or scprod. Z  is either sqdistmatrix or scprodmatrix and represents matrix of pair-wise z(x,y) values for two sets of vectors X and Y (previously gramian and lagged gramian).

So basically, K(X,Y) = elementwise_kappa(Z(X,Y)) where Z() depends on the class of kernel and kappa() depends on the kernel itself. This way kappa (and its derivatives) can be defined once and are usually one-liners. Also, if we ever need to compute K1(X,Y) and K2(X,Y) and k1 and k2 are kernels of the same class, then they can reuse the initial Z(X,Y) matrix.

I'd like to take a similar approach to derivatives. I tried playing around with computing that initial reusable matrix equivalent for kernel derivatives, but there's something weird with my code and it's going slow right now, but I haven't found out where the performance gap is. It's called 'tensor_epsilons'.
